subsys/caf/modules/sensor_data_aggregator.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <drivers/sensor.h>
+#include <event_manager.h>
+
+#include <caf/events/sensor_event.h>
+#include <caf/events/sensor_aggregator_event.h>
+#include <caf/sensor_manager.h>
+
+#define MODULE sensor_data_aggregator
+#include <caf/events/module_state_event.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_CAF_SENSOR_MANAGER_LOG_LEVEL);
+
+#define DT_DRV_COMPAT caf_aggregator
+
+#define __DEFINE_DATA(i, agg_id, size)							\
+	__ALIGNED(4) static uint8_t agg_ ## agg_id ## _buff_ ## i ## _data[size];
+
+#define __INITIALIZE_BUF(i, agg_id)							\
+	{(uint8_t *) &agg_ ## agg_id ## _buff_ ## i ## _data				\
+	},
+
+/* buf_len has to be equal to n*sensor_data_size*/
+#define __DEFINE_BUF_DATA(i)									\
+	UTIL_LISTIFY(DT_INST_PROP(i, buf_count), __DEFINE_DATA, i,				\
+		DT_INST_PROP(i, buf_data_length))						\
+	static struct aggregator_buffer agg_ ## i ## _bufs[] = {				\
+	UTIL_LISTIFY(DT_INST_PROP(i, buf_count), __INITIALIZE_BUF, i)				\
+	};											\
+	BUILD_ASSERT((DT_INST_PROP(i, buf_data_length) % DT_INST_PROP(i, sensor_data_size)) == 0,\
+		"Wrong sensor data or buffer size");
+
+
+#define __DEFINE_AGGREGATOR(i)								\
+	[i].sensor_descr = DT_INST_PROP(i, sensor_descr),				\
+	[i].sensor_data_size = DT_INST_PROP(i, sensor_data_size),			\
+	[i].buf_count = DT_INST_PROP(i, buf_count),					\
+	[i].buf_len = DT_INST_PROP(i, buf_data_length),					\
+	[i].agg_buffers = (struct aggregator_buffer *) &agg_ ## i ## _bufs,		\
+	[i].active_buf = (struct aggregator_buffer *) &agg_ ## i ## _bufs,
+
+#define __DEFINE_AGG							\
+	static struct aggregator aggregators[] = {			\
+	DT_INST_FOREACH_STATUS_OKAY(__DEFINE_AGGREGATOR)		\
+	}
+
+struct aggregator_buffer {
+	uint8_t *data;		/* Dynamic data. */
+	bool busy;		/* Buffer status. */
+	uint8_t sample_cnt;	/* Number of samples already saved in the buffer. */
+};
+
+struct aggregator {
+	const char *sensor_descr;			/* sensor_description of the sensor. */
+	struct aggregator_buffer *agg_buffers;	/* Buffers */
+	struct aggregator_buffer *active_buf;	/* Active buffer to which data will be placed */
+	enum sensor_state sensor_state;		/* Sensors state */
+	const uint8_t sensor_data_size;		/* Size of sensor data in bytes. */
+	const uint8_t buf_count;		/* Number of buffers */
+	const uint8_t buf_len;			/* Size of buffor data in bytes. */
+};
+
+
+DT_INST_FOREACH_STATUS_OKAY(__DEFINE_BUF_DATA) //no semicolon on purpose.
+__DEFINE_AGG;
```

Comment:
We don't need this marco - let's use content directly.

Justification - the content is here anyway and is used only once. By adding `__DEFINE_AGG` you add additional lines and one more entity to track.